### ISSUE TITLE
Mounting config map into CRD

### ISSUE THREAD

**Issue by roksolana-d** [fabric8io/kubernetes-client#1537]
I have a CRD and config map that I created. Is this possible to mount a config map into a custom object?

Here is the way I define a custom object:

  ```
  val sparkCustomResource = client.customResource(crdContext)
    sparkCustomResource.create("default", loadYaml(sparkAppFile))

    val configMapResource = client.configMaps()
      .inNamespace("default")
      .withName("app-config")

    val configMap: ConfigMap = configMapResource.createOrReplace(new ConfigMapBuilder()
      .withNewMetadata()
      .withName("app-config")
      .endMetadata()
      .addToData("application-conf", processStringFromHocon(confFilePath))
    .build())

```
The loadYaml method reads yaml file and loads it as an InputSream, processStringFromHocon method - reads application.conf file into String.

**Comment by rohanKanojia** [comment]
I haven't tried it. Are you able to do it via `kubectl`? Can you share your Custom Resource Definition so that I can see how a custom resource is mounted(it can be a list like [here](https://github.com/coreos/prometheus-operator/blob/master/example/prometheus-operator-crd/prometheus.crd.yaml#L800)?

**Comment by roksolana-d** [comment]
I did not define volume mounting in CRD as I would like to do this using library functionality.
My CRD looks like this:

```
apiVersion: sparkoperator.k8s.io/v1beta1
kind: SparkApplication
metadata:
  name: spark-mssql-job
  namespace: default
spec:
  type: Scala
  image: gcr.io/spark-operator/spark:v2.4.0-gcs-prometheus
  imagePullPolicy: Always
  mainClass: mssql_job.SparkMssqlJob
  mainApplicationFile: jar_name.jar
  mode: cluster
  driver:
    coreLimit: 1000m
    cores: 0.1
    labels:
      version: 2.4.0
    memory: 1024m
    serviceAccount: default
  executor:
    cores: 1
    instances: 1
    labels:
      version: 2.4.0
    memory: 1024m
  monitoring:
    exposeDriverMetrics: true
    exposeExecutorMetrics: true
    prometheus:
      jmxExporterJar: "/prometheus/jmx_prometheus_javaagent-0.3.1.jar"
      port: 8090
```

**Comment by rohanKanojia** [comment]
I'm afraid this kind of functionality is not supported right now. Just to understand, how do you mount a configmap using `kubectl`? Does it provide some kind of mount utility that does so? I thought maybe we have to edit CR's spec in order to mount.

**Comment by roksolana-d** [comment]
I would just specify a volume to mount in a custom object definition. I do not know any other utilities to do so.

**Comment by rohanKanojia** [comment]
Sorry, It's still not so clear to me. Can you share an example with me? Maybe modified spec yaml or some link where it's shown.

**Comment by roksolana-d** [comment]
The example of config map would look like this:

```
apiVersion: sparkoperator.k8s.io/v1beta1
kind: SparkApplication
metadata:
  name: spark-mssql-job
  namespace: default
spec:
  type: Scala
  image: gcr.io/spark-operator/spark:v2.4.0-gcs-prometheus
  imagePullPolicy: Always
  mainClass: mssql_job.SparkMssqlJob
  mainApplicationFile: local:///tmp/spark_job.jar
  mode: cluster
  driver:
    coreLimit: 1000m
    cores: 0.1
    labels:
      version: 2.4.0
    memory: 1024m
    serviceAccount: default
    volumeMounts:
      - name: app-config
        mountPath: /tmp
  executor:
    cores: 1
    instances: 1
    labels:
      version: 2.4.0
    memory: 1024m
    volumeMounts:
      - name: app-config
        mountPath: /tmp
  monitoring:
    exposeDriverMetrics: true
    exposeExecutorMetrics: true
    prometheus:
      jmxExporterJar: "/prometheus/jmx_prometheus_javaagent-0.3.1.jar"
      port: 8090
  volumes:
    - name: app-config
      configMap:
        name: app-conf
```

**Comment by rohanKanojia** [comment]
@roksolana-d : Does this work for you? 
```
         Map<String, Object> configMapVolume = new HashMap<>();
         configMapVolume.put("name", "app-config");
         configMapVolume.put("configMap", Collections.singletonMap("name", "app-conf"));

         Map<String, Object> spartCustomResourceSpec = (Map<String, Object>)sparkCustomResource.get("spec");
         sparkCustomResourceSpec.put("volumes", Arrays.asList(configMapVolume));
```

**Comment by roksolana-d** [comment]
I get this exception on custom resource get spec call:

```
t_tests.SparkMssqlJobIntegrationTest *** ABORTED ***
[info]   java.lang.IllegalStateException: Bad Request
[info]   at io.fabric8.kubernetes.client.dsl.internal.RawCustomResourceOperationsImpl.makeCall(RawCustomResourceOperationsImpl.java:150)
[info]   at io.fabric8.kubernetes.client.dsl.internal.RawCustomResourceOperationsImpl.get(RawCustomResourceOperationsImpl.java:79)

```

I think it should be changed another way as sparkCustomResources is more like a custom resource while it is needed to modify yaml file of the custom object.

**Comment by roksolana-d** [comment]
@rohanKanojia can you help with this issue?

**Comment by rohanKanojia** [comment]
Hi, will try to try it myself locally. Otherwise we would have to add support for this.

**Comment by rohanKanojia** [comment]
I tried applying your resource onto cluster using plain `kubectl` but it seems to be failing:
```
~/work/k8-resource-yamls : $ ls
sparkapplication-crd.yml  sparkapplication-cr.yml
~/work/k8-resource-yamls : $ kubectl get crds
NAME                                     CREATED AT
sparkapplications.sparkoperator.k8s.io   2019-06-20T12:02:33Z
~/work/k8-resource-yamls : $ kubectl apply -f sparkapplication-cr.yml 
The SparkApplication "spark-mssql-job" is invalid: []: Invalid value: map[string]interface {}{"apiVersion":"sparkoperator.k8s.io/v1beta1", "kind":"SparkApplication", "metadata":map[string]interface {}{"annotations":map[string]interface {}{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"sparkoperator.k8s.io/v1beta1\",\"kind\":\"SparkApplication\",\"metadata\":{\"annotations\":{},\"name\":\"spark-mssql-job\",\"namespace\":\"default\"},\"spec\":{\"driver\":{\"coreLimit\":\"1000m\",\"cores\":0.1,\"labels\":{\"version\":\"2.4.0\"},\"memory\":\"1024m\",\"serviceAccount\":\"default\",\"volumeMounts\":[{\"mountPath\":\"/tmp\",\"name\":\"app-config\"}]},\"executor\":{\"cores\":1,\"instances\":1,\"labels\":{\"version\":\"2.4.0\"},\"memory\":\"1024m\",\"volumeMounts\":[{\"mountPath\":\"/tmp\",\"name\":\"app-config\"}]},\"image\":\"gcr.io/spark-operator/spark:v2.4.0-gcs-prometheus\",\"imagePullPolicy\":\"Always\",\"mainApplicationFile\":\"local:///tmp/spark_job.jar\",\"mainClass\":\"mssql_job.SparkMssqlJob\",\"mode\":\"cluster\",\"monitoring\":{\"exposeDriverMetrics\":true,\"exposeExecutorMetrics\":true,\"prometheus\":{\"jmxExporterJar\":\"/prometheus/jmx_prometheus_javaagent-0.3.1.jar\",\"port\":8090}},\"type\":\"Scala\"}}\n"}, "creationTimestamp":"2019-06-20T12:13:12Z", "generation":1, "name":"spark-mssql-job", "namespace":"default", "uid":"c62f59a3-9354-11e9-b4dd-0800278bfe0a"}, "spec":map[string]interface {}{"driver":map[string]interface {}{"coreLimit":"1000m", "cores":0.1, "labels":map[string]interface {}{"version":"2.4.0"}, "memory":"1024m", "serviceAccount":"default", "volumeMounts":[]interface {}{map[string]interface {}{"mountPath":"/tmp", "name":"app-config"}}}, "executor":map[string]interface {}{"cores":1, "instances":1, "labels":map[string]interface {}{"version":"2.4.0"}, "memory":"1024m", "volumeMounts":[]interface {}{map[string]interface {}{"mountPath":"/tmp", "name":"app-config"}}}, "image":"gcr.io/spark-operator/spark:v2.4.0-gcs-prometheus", "imagePullPolicy":"Always", "mainApplicationFile":"local:///tmp/spark_job.jar", "mainClass":"mssql_job.SparkMssqlJob", "mode":"cluster", "monitoring":map[string]interface {}{"exposeDriverMetrics":true, "exposeExecutorMetrics":true, "prometheus":map[string]interface {}{"jmxExporterJar":"/prometheus/jmx_prometheus_javaagent-0.3.1.jar", "port":8090}}, "type":"Scala"}}: validation failure list:
spec.sparkVersion in body is required
~/work/k8-resource-yamls : $ 
```

Are you sure the cr manifest is correct? Or maybe I missed something here?

**Comment by rohanKanojia** [comment]
Hm, I added `sparkVersion: 2.4.3` and it got applied. Sorry for the noise :-|

**Comment by rohanKanojia** [comment]
I started hitting #1565 when I create spark crd. Maybe that's why you were unable to update?

**Comment by rohanKanojia** [comment]
@roksolana-d : I've created #1600 to fix #1565 , when I tried editing spark custom resource for configmap. I was able to do it like this:

```
        try (final KubernetesClient client = new DefaultKubernetesClient()) {
            CustomResourceDefinition sparkCrd = client.customResourceDefinitions()
                    .load(SparkApplicationCustomResource.class.getResourceAsStream("/sparkapplication-crd.yml"))
                    .get();
            client.customResourceDefinitions().create(sparkCrd);
            log("Custom Resource Definition " + sparkCrd.getMetadata().getName() + " created.");

            CustomResourceDefinitionContext customResourceDefinitionContext = new CustomResourceDefinitionContext.Builder()
                    .withGroup("sparkoperator.k8s.io")
                    .withName("sparkapplications.sparkoperator.k8s.io")
                    .withPlural("sparkapplications")
                    .withScope("Namespaced")
                    .withVersion("v1beta1")
                    .build();

            // Create Spark Application Custom Resource
            Map<String, Object> sparkCustomResource = client.customResource(customResourceDefinitionContext)
                    .create("default", SparkApplicationCustomResource.class.getResourceAsStream("/sparkapplication-cr.yml"));
            System.out.println(sparkCustomResource.toString());

            String customResourceName = ((Map<String, Object> )sparkCustomResource.get("metadata")).get("name").toString();
            log("Custom resource " + customResourceName + " created");

            // Add volume
            Map<String, Object> configMapVolume = new HashMap<>();
            configMapVolume.put("name", "app-config");
            configMapVolume.put("configMap", Collections.singletonMap("name", "app-conf"));
            Map<String, Object> sparkCustomResourceSpec = (Map<String, Object>)sparkCustomResource.get("spec");
            sparkCustomResourceSpec.put("volumes", Arrays.asList(configMapVolume));

            sparkCustomResource.put("spec", sparkCustomResourceSpec);

            client.customResource(customResourceDefinitionContext).edit("default", customResourceName, sparkCustomResource);
            log("Edition of custom resource successful.");

        }
```
I verified it after running this code:
```
~/go/src/github.com/fabric8io/kubernetes-client : $ kubectl get crds
NAME                                     CREATED AT
sparkapplications.sparkoperator.k8s.io   2019-06-22T06:41:43Z
~/go/src/github.com/fabric8io/kubernetes-client : $ kubectl get sparkapplications.sparkoperator.k8s.io
NAME              AGE
spark-mssql-job   11s
~/go/src/github.com/fabric8io/kubernetes-client : $ kubectl get spark-mssql-job -o yaml
error: the server doesn't have a resource type "spark-mssql-job"
~/go/src/github.com/fabric8io/kubernetes-client : $ kubectl get sparkapplication spark-mssql-job -o yaml
apiVersion: sparkoperator.k8s.io/v1beta1
kind: SparkApplication
metadata:
  creationTimestamp: "2019-06-22T06:41:43Z"
  generation: 2
  name: spark-mssql-job
  namespace: default
  resourceVersion: "28851"
  selfLink: /apis/sparkoperator.k8s.io/v1beta1/namespaces/default/sparkapplications/spark-mssql-job
  uid: cca3e43b-94b8-11e9-9b0f-0800278bfe0a
spec:
  driver:
    coreLimit: 1000m
    cores: 0.1
    labels:
      version: 2.4.0
    memory: 1024m
    serviceAccount: default
    volumeMounts:
    - mountPath: /tmp
      name: app-config
  executor:
    cores: 1
    instances: 1
    labels:
      version: 2.4.0
    memory: 1024m
    volumeMounts:
    - mountPath: /tmp
      name: app-config
  image: gcr.io/spark-operator/spark:v2.4.0-gcs-prometheus
  imagePullPolicy: Always
  mainApplicationFile: local:///tmp/spark_job.jar
  mainClass: mssql_job.SparkMssqlJob
  mode: cluster
  monitoring:
    exposeDriverMetrics: true
    exposeExecutorMetrics: true
    prometheus:
      jmxExporterJar: /prometheus/jmx_prometheus_javaagent-0.3.1.jar
      port: 8090
  sparkVersion: 2.4.3
  type: Scala
  volumes:
  - configMap:
      name: app-conf
    name: app-config
~/go/src/github.com/fabric8io/kubernetes-client : $ 
```

You can try out snapshot build of #1600 or wait until next release.

**Comment by rohanKanojia** [comment]
@roksolana-d : Hi, Sorry for the delay. Could you please try this again with v4.3.1?
